packages/ics721-types/src/token_types.rs

```diff
@@ -0,0 +1,145 @@
+use std::ops::Deref;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Binary, StdResult};
+use cw_storage_plus::{Bound, Bounder, Key, KeyDeserialize, Prefixer, PrimaryKey};
+
+/// A token ID according to the ICS-721 spec. The newtype pattern is
+/// used here to provide some distinction between token and class IDs
+/// in the type system.
+#[cw_serde]
+pub struct TokenId(String);
+
+/// A token according to the ICS-721 spec.
+#[cw_serde]
+pub struct Token {
+    /// A unique identifier for the token.
+    pub id: TokenId,
+    /// Optional URI pointing to off-chain metadata about the token.
+    pub uri: Option<String>,
+    /// Optional base64 encoded metadata about the token.
+    pub data: Option<Binary>,
+}
+
+/// A class ID according to the ICS-721 spec. The newtype pattern is
+/// used here to provide some distinction between token and class IDs
+/// in the type system.
+#[cw_serde]
+pub struct ClassId(String);
+
+#[cw_serde]
+pub struct Class {
+    /// A unique (from the source chain's perspective) identifier for
+    /// the class.
+    pub id: ClassId,
+    /// Optional URI pointing to off-chain metadata about the class.
+    pub uri: Option<String>,
+    /// Optional base64 encoded metadata about the class.
+    pub data: Option<Binary>,
+}
+
+impl TokenId {
+    pub fn new<T>(token_id: T) -> Self
+    where
+        T: Into<String>,
+    {
+        Self(token_id.into())
+    }
+}
+
+impl ClassId {
+    pub fn new<T>(class_id: T) -> Self
+    where
+        T: Into<String>,
+    {
+        Self(class_id.into())
+    }
+}
+
+impl<'a> Bounder<'a> for ClassId {
+    fn inclusive_bound(self) -> Option<cw_storage_plus::Bound<'a, Self>> {
+        Some(Bound::inclusive(self))
+    }
+
+    fn exclusive_bound(self) -> Option<cw_storage_plus::Bound<'a, Self>> {
+        Some(Bound::exclusive(self))
+    }
+}
+
+// boilerplate for conversion between wrappers and the wrapped.
+
+impl From<ClassId> for String {
```

Comment:
Why do we need all these boiler plate functions?

Reply:
All these code in `ics721-types` exists long ago and have been provided by Zeke.
ClassId is a string, but explicitely defined as a type for better readability. Otherwise there would be for Class, ClassId, TokenId -> String, String, and String - which isnt really helpful :).

Implementing From trait allows types like ClassId automatically converted/inferred into String without explicitly calling `to_string()`

Reply:
Well then its not "boiler plate" no? Its convenience functions for functionality on "ClassId" which is really just a string but we're making it a type by giving it these functions.

Reply:
Done